riscv/mmu.cc

```diff
@@ -190,7 +190,11 @@ void mmu_t::load_slow_path_intrapage(reg_t addr, reg_t len, uint8_t* bytes, uint
       tracer.trace(paddr, len, LOAD);
     else if (xlate_flags == 0)
       refill_tlb(addr, paddr, host_addr, LOAD);
-  } else if (!mmio_load(paddr, len, bytes)) {
+
+    if (xlate_flags & RISCV_XLATE_LR) {
+      load_reservation_address = paddr;
+    }
+  } else if ((xlate_flags & RISCV_XLATE_LR) || !mmio_load(paddr, len, bytes)) {
     throw trap_load_access_fault((proc) ? proc->state.v : false, addr, 0, 0);
```

Comment:
While LR operation cannot access mmio, how about amo operation?
If it can, how to handle the cases when the mmio address is read-only or write-only?

Reply:
(Note this PR makes no functional change with respect to AMOs on MMIO regions.)

Reply:
Before performing the read, write permissions haven't been checked for MMIOs currently. So the actual store may fail after the load success.

Reply:
> Note this PR does not change any policy re: AMO/LR/SC to MMIO. It only fixes the LR precise exception bug. Otherwise, the exception behavior for all of these operations is unchanged.

Yeah. I know. I just thought of this.

Reply:
> Before performing the read, write permissions haven't been checked for MMIOs currently. So the actual store may fail after the load success.

This might be acceptable for MMIO, as often this kind of thing is dealt with using (imprecise) bus errors.  We should always be precise when possible, but some amount of imprecision is inevitable.